### PR TITLE
fix(core): prevent NgModule scope being overwritten in JIT compiler

### LINKED ISSUE #37105: pipe not found in ng serve but found in production build
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [ ] build
- [x] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] config
- [ ] run
- [ ] help
- [ ] version
- [ ] doc

### Description
i try to upgrade from angular 8 to angular 9. i found one of my component can not find any pipe. but other component in same module works. and in production build every component can find a pipe normally.

## 🔬 Minimal Reproduction
https://github.com/mix5003/ng-pipe-not-exists
this is from my project that remove not relate code already . if you run `ng serve` and go to `/2/report/financial/profit-and-loss` it works. but if you go to `/2/report/financial/health-index` it will throw error. 

if you try to copy all code from `FinancialBaseComponent` to `FinancialBusinessHealthComponent` and remove `extends FinancialBaseComponent` it works too.

and if you create new project and copy `src/app` to new project and run `ng serve` it works too. 

## 🔥 Exception or Error
<pre><code>
core.js:4098 ERROR Error: Uncaught (in promise): Error: The pipe 'translate' could not be found!
Error: The pipe 'translate' could not be found!
    at getPipeDef$1 (core.js:25434)
    at ɵɵpipe (core.js:25397)
    at FinancialBusinessHealthComponent_Template (VM89 FinancialBusinessHealthComponent.js:20)
    at executeTemplate (core.js:7706)
    at renderView (core.js:7523)
    at renderComponent (core.js:8747)
    at renderChildComponents (core.js:7382)
    at renderView (core.js:7548)
    at ComponentFactory.push../node_modules/@angular/core/__ivy_ngcc__/fesm5/core.js.ComponentFactory.create (core.js:23305)
    at ViewContainerRef.createComponent (core.js:10531)
    at resolvePromise (zone.js:832)
    at resolvePromise (zone.js:784)
    at zone.js:894
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:421)
    at Object.onInvokeTask (core.js:28273)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:420)
    at Zone.push../node_modules/zone.js/dist/zone.js.Zone.runTask (zone.js:188)
    at drainMicroTaskQueue (zone.js:601)
    at ZoneTask.push../node_modules/zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (zone.js:507)
    at invokeTask (zone.js:1671)
</code></pre>

## 🌍 Your Environment
<pre><code>
Angular CLI: 9.1.4
Node: 12.16.2
OS: win32 x64

Angular: 9.1.4
... animations, cli, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router
Ivy Workspace: <error>

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.901.4
@angular-devkit/build-angular     0.901.4
@angular-devkit/build-optimizer   0.901.4
@angular-devkit/build-webpack     0.901.4
@angular-devkit/core              9.1.4
@angular-devkit/schematics        9.1.4
@ngtools/webpack                  9.1.4
@schematics/angular               9.1.4
@schematics/update                0.901.4
rxjs                              6.5.5
typescript                        3.8.3
webpack                           4.42.0
</code></pre>

**Anything else relevant?**
if you think this is not bug feel free to close this issue. i just curious why it happen?

### PR DESCRIPTION
**Pull request by JoostK**
In JIT compiled apps, component definitions are compiled upon first
access. For a component class `A` that extends component class `B`, the
`B` component is also compiled when the `InheritDefinitionFeature` runs
during the compilation of `A` before it has finalized. A problem arises
when the compilation of `B` would flush the NgModule scoping queue,
where the NgModule declaring `A` is still pending. The scope information
would be applied to the definition of `A`, but its compilation is still
in progress so requesting the component definition would compile `A`
again from scratch. This "inner compilation" is correctly assigned the
NgModule scope, but once the "outer compilation" of `A` finishes it
would overwrite the inner compilation's definition, losing the NgModule
scope information.

In summary, flushing the NgModule scope queue could trigger a reentrant
compilation, where JIT compilation is non-reentrant. To avoid the
reentrant compilation, a compilation depth counter is introduced to
avoid flushing the NgModule scope during nested compilations.

Fixes #37105

### CODE REVIEW

File: packages/core/src/render3/jit/directive.ts [hunk 1]
```diff
@@ -106,18 +119,25 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
           interpolation: metadata.interpolation,
           viewProviders: metadata.viewProviders || null,
         };
-        if (meta.usesInheritance) {
-          addDirectiveDefToUndecoratedParents(type);
-        }
 
-        ngComponentDef = compiler.compileComponent(angularCoreEnv, templateUrl, meta);
+        compilationDepth++;
+        try {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/jit/directive.ts:124]: @JoostK thanks for the fix 👍 

May be add a quick comment on why try-catch was added here?

**Review by alxhub — APPROVED** [review]

**Comment by alxhub** [comment]
[Presubmit](http://test/OCL:320636464:BASE:320636519:1594404462229:31e0532a)

**Comment by alxhub** [comment]
[And again](http://test/OCL:320636464:BASE:324706035:1596496768034:f88f2b68)
